include/range/v3/utility/polymorphic_cast.hpp

```diff
@@ -5,18 +5,31 @@
 
 #ifndef RANGES_V3_UTILITY_POLYMORPHIC_CAST_HPP
 #define RANGES_V3_UTILITY_POLYMORPHIC_CAST_HPP
+
+#include <memory>
+#include <type_traits>
+#include <meta/meta.hpp>
 #include <range/v3/detail/config.hpp>
 
 namespace ranges
 {
     inline namespace v3
     {
         template<typename Target, typename Source>
-        inline Target polymorphic_downcast(Source* x)
+        meta::if_<std::is_pointer<Target>, Target>
+        polymorphic_downcast(Source *x) noexcept
         {
             RANGES_ASSERT(dynamic_cast<Target>(x) == x);
             return static_cast<Target>(x);
         }
+        template<typename Target, typename Source>
+        meta::if_<std::is_reference<Target>, Target>
+        polymorphic_downcast(Source &&x) noexcept
+        {
+            using PTarget = meta::_t<std::add_pointer<Target>>;
+            auto ptr = polymorphic_downcast<PTarget>(std::addressof(x));
```

Comment:
Well, I suppose someone might want to `polymorphic_downcast<Derived&&>(std::move(base))`. But I would prefer to write `std::move(polymorphic_downcast<Derived&>(base))`, and have a safer `polymorphic_downcast`.